server/use_cases/add_service_commitments.py

```diff
@@ -39,52 +41,98 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
+    final_valid_commitments = []
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        # Check for overlap with existing user shifts
+        for existing_shift in existing_user_shifts:
+            if (current_shift.shift_start < existing_shift.shift_end and
+                current_shift.shift_end > existing_shift.shift_start):
+                has_overlap = True
+                break
+        # Check for overlap with previously approved shifts in this request
+        if not has_overlap:
+            approved_shifts = [
+                valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+            for approved_shift in approved_shifts:
+                if (current_shift.shift_start < approved_shift.shift_end and
+                    current_shift.shift_end > approved_shift.shift_start):
+                    has_overlap = True
+                    break
+        if has_overlap:
+            results[idx] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Overlapping commitment"
             }
+        else:
+            final_valid_commitments.append(valid_commitments[i])
+            final_valid_indexes.append(idx)
+            # Add current shift to existing_user_shifts
+            # to check against subsequent commitments
+            existing_user_shifts.append(current_shift)
+    if not final_valid_commitments:
         return results
-
-    # insert valid commitments
-    commitments_as_dict = [c.to_dict() for c in valid_commitments]
-    inserted_commitments = (
-        commitments_repo.insert_service_commitments(
-            (commitments_as_dict)
-        )
-    )
-
-    # fill in success results for valid commitments
-    for idx, i in enumerate(valid_indexes):
+    commitments_as_dict = [c.to_dict() for c in final_valid_commitments]
+    inserted_commitments = commitments_repo.insert_service_commitments(
+        commitments_as_dict)
+    for idx, i in enumerate(final_valid_indexes):
         if idx < len(inserted_commitments):
             results[i] = {
-                "service_commitment_id": 
-                    str(inserted_commitments[idx]),
+                "service_commitment_id": str(inserted_commitments[idx]),
                 "success": True
             }
         else:
             results[i] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Failed to insert commitment"
+            }
+    for i in valid_indexes:
+        if results[i] is None:
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
             }
-
     return results
 
-
 def check_time_overlap(shifts):
```

Comment:
You should change this function to be:
```suggestion
def check_time_overlap(shift_to_check, other_shifts):
```
Have this function return true if shift_to_check overlaps with any of the shifts in other_sifts.
Then use this function in your code